src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.cs

```diff
@@ -23,65 +25,242 @@ namespace System.Text.RegularExpressions.Generator
+                    // We'll still output a limited implementation that just caches a new Regex(...).
+                    if (!regexType.Method.Tree.Root.SupportsCompilation(out string? reason))
+                    {
+                        return (regexType, reason, Diagnostic.Create(DiagnosticDescriptors.LimitedSourceGeneration, regexType.Method.MethodSyntax.GetLocation()));
+                    }
+
+                    // Generate the core logic for the regex.
+                    Dictionary<string, string[]> requiredHelpers = new();
+                    var sw = new StringWriter();
+                    var writer = new IndentedTextWriter(sw);
+                    writer.Indent += 3;
+                    writer.WriteLine();
+                    EmitRegexDerivedTypeRunnerFactory(writer, regexType.Method, requiredHelpers);
+                    writer.Indent -= 3;
+                    return (regexType, sw.ToString(), requiredHelpers);
                 })
                 .Collect();
 
+            // To avoid invalidating every regex's output when anything from the compilation changes,
+            // we extract from it the only things we care about: whether unsafe code is allowed,
+            // and a name based on the assembly's name, and only that information is then fed into
+            // RegisterSourceOutput along with all of the cached generated data from each regex.
+            IncrementalValueProvider<(bool AllowUnsafe, string? AssemblyName)> compilationDataProvider =
+                context.CompilationProvider
+                .Select((x, _) => (x.Options is CSharpCompilationOptions { AllowUnsafe: true }, x.AssemblyName));
+
             // When there something to output, take all the generated strings and concatenate them to output,
             // and raise all of the created diagnostics.
-            context.RegisterSourceOutput(codeOrDiagnostics, static (context, results) =>
+            context.RegisterSourceOutput(codeOrDiagnostics.Combine(compilationDataProvider), static (context, compilationDataAndResults) =>
             {
-                var code = new List<string>(s_headers.Length + results.Length);
+                ImmutableArray<object?> results = compilationDataAndResults.Left;
+
+                // Report any top-level diagnostics.
+                bool allFailures = true;
+                foreach (object? result in results)
+                {
+                    if (result is Diagnostic d)
+                    {
+                        context.ReportDiagnostic(d);
+                    }
+                    else
+                    {
+                        allFailures = false;
+                    }
+                }
+                if (allFailures)
+                {
+                    return;
+                }
+
+                // At this point we'll be emitting code.  Create a writer to hold it all.
+                var sw = new StringWriter();
```

Comment:
Not really very important here, but would it make it cleaner if we moved all of the code being written here under RegexGenerator.Emitter.cs instead? I kind of liked the separation we had where RegexGenerator.cs was the main hook and initialization piece for the generator, parser would be the thing that detects methods to generate code for, and emitter for the actual writing of the code. With this PR, a big portion of code emitting is happening in RegexGenerator.cs which I understand is mostly just boilerplate but might be cleaner to move to Emitter. No need to do it on this PR, we can do it later.